### PR TITLE
[IMP] add history in CoreView plugins

### PR DESCRIPTION
**Pull request by hokolomopo**
## Description:

description of this task, what is implemented and why it is implemented that way.

Odoo task ID : [3288703](https://www.odoo.com/web#id=3288703&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

**Review by LucasLefevre — COMMENTED** [review]
:+1: 
Only two comments

**Comment by LucasLefevre** [comment]
Also, we could maybe go further: do core views plugins currently need UI getters ? If not, we could just give them core getters to ensure it remains that way.

File: src/plugins/ui_core_views/custom_colors.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ import {
   toHex,
 } from "../../helpers";
 import { GaugeChart, ScorecardChart } from "../../helpers/figures/charts";
-import { Cell, Color, CoreViewCommand, RGBA, UID } from "../../types";
+import { Cell, Color, CoreViewCommand, Immutable, RGBA, UID } from "../../types";
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/ui_core_views/custom_colors.ts:13]: 'cause it tanks the performances for not much gain (evaluation for example)

File: src/model.ts [hunk 2]
```diff
@@ -366,6 +366,7 @@ export class Model extends EventBus<any> implements CommandDispatcher {
       this.dispatchFromCorePlugin("REDO", { commands });
       this.finalize();
     });
+    this.session.on("snapshot", this, this.finalize);
```
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply src/model.ts:364]: > it's more correct to still add it

Not really, it isn't supposed to do/change anything

File: src/plugins/ui_plugin.ts [hunk 3]
```diff
@@ -25,8 +25,19 @@ export interface UIPluginConfig {
   readonly session: Session;
 }
 
-export interface UIPluginConstructor {
-  new (config: UIPluginConfig): UIPlugin;
+export interface CoreViewPluginConfig extends CommonUIPluginConfig {
+  readonly getters: CoreViewGetters;
+  readonly allGetters: Getters;
```
Comment (hokolomopo) [inline src/plugins/ui_plugin.ts:30]: EvaluationPlugin needs all getters to give to the functions for their evaluation (eg. odoo list functions)  :x

File: src/index.ts [hunk 4]
```diff
@@ -110,7 +110,7 @@ export { findCellInNewZone } from "./helpers/zones";
 export { load } from "./migrations/data";
 export { Model } from "./model";
 export { CorePlugin } from "./plugins/core_plugin";
-export { UIPlugin } from "./plugins/ui_plugin";
+export { BasicUIPlugin, CoreViewPlugin } from "./plugins/ui_plugin";
```
Comment (hokolomopo) [inline src/index.ts:113]: Need odoo adaptation if we keep this commit

**Review by hokolomopo — COMMENTED** [review]
@LucasLefevre added a commit to only give coreView getters to CoreView plugins + typing. Check if you like it. Evaluation makes it a bit uglier that it should be.

**Comment by LucasLefevre** [comment]
> @LucasLefevre I added a commit to only give coreView getters to CoreView plugins + typing. Check if you like it. Evaluation makes it a bit uglier that it should be.

Thanks !
Hmm maybe it's worth keeping things simple.

File: src/types/misc.ts [hunk 5]
```diff
@@ -348,3 +348,18 @@ export interface SortOptions {
 export interface MenuMouseEvent extends MouseEvent {
   closedMenuId?: UID;
 }
+
+// https://github.com/Microsoft/TypeScript/issues/13923#issuecomment-557509399
+// prettier-ignore
+export type Immutable<T> =
```
Comment (rrahir) [inline src/types/misc.ts:354]: deserves a dedicated commit imho, it's really nice but people should be aware of its existence to somehow use it

File: src/plugins/ui_core_views/evaluation_conditional_format.ts [hunk 6]
```diff
@@ -176,7 +174,7 @@ export class EvaluationConditionalFormatPlugin extends UIPlugin {
       case "percentile":
         return percentile(rangeValues, Number(threshold.value) / 100, true);
       case "formula":
-        const value = threshold.value && this.getters.evaluateFormula(threshold.value);
+        const value = threshold.value && this.getters.evaluateFormula(sheetId, threshold.value);
         return !(value instanceof Promise) ? value : null;
```
Comment (rrahir) [inline src/plugins/ui_core_views/evaluation_conditional_format.ts:178]: do we actually support this feature ? :p

**Review by LucasLefevre — APPROVED** [review]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo rebase-ff
